src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -1078,7 +1078,12 @@ private void AppendWithExpansion(char value)
             return Append(value.ToString(format, provider));
         }
 
-        public StringBuilder Append(object? value) => (value == null) ? this : Append(value.ToString());
+        public StringBuilder Append(object? value) => value switch
+        {
+            ISpanFormattable sf => AppendSpanFormattable(sf),
```

Comment:
If I'm reading this file correctly, the `AppendSpanFormattable` method is not designed for arbitrary implementations of `ISpanFormattable`. It has this assert:

```c#
Debug.Assert(typeof(T).Assembly.Equals(typeof(object).Assembly), "Implementation trusts the results of TryFormat because T is expected to be something known");
```

However since `T` will be `ISpanFormattable` when called from here, it will always pass, even on types defined outside System.Private.CoreLib. If this assert is doing something useful, it should probably use `value.GetType()` instead of `typeof(T)` .

Reply:
I think this `AppendFormatted_InvalidTryFormatCharsWritten_Throws` method shows what sort of extra validation needs to be done:

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/tests/System/Text/StringBuilderInterpolationTests.cs#L437

A `FormatException` should be thrown if the `charsWritten` value from `TryFormat` is larger than the destination.

Reply:
You would also need to make sure `charsWritten` isn't returning a negative value now when user defined `ISpanFormattable` is allowed, to make sure m_ChunkLength remains guarded

Reply:
> isn't returning a negative value

Negative or larger than the available space (the cited line addresses both). Other code elsewhere in StringBuilder uses unsafe code, and we need to be sure we're not leaving the state of the instance here in a way that will cause that unsafe code to start reading/writing over unowned memory.